### PR TITLE
Add Fill & Outline styles, width settings & new typography controls for Product(Add to cart) button block

### PR DESCRIPTION
**Pull request by imanish003**
This PR is part of #8085 & adds following controls to product(Add to cart) button in Inspector controls.
#### Styles: Fill & Outline
Here is a quick demo video:

https://user-images.githubusercontent.com/16707866/226090143-417c53ad-8de5-4ed4-81c9-ec9628e64471.mov

#### Width Settings
The button block now has a width selector, which allows the user to set the button to 25%, 50%, 75%, or 100% of the parent container. By default, a button's width is determined by the size of its content.

You can find this setting in the button block's sidebar.
image

#### Typography controls

Now following typography controls are also available:
- Font Family
- Appearance
- Line height
- Letter spacing
- Decoration
- Letter case

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

- Create a new post & add `Products block` to it.
- In products block, you can find `Add to cart button` as shown in screenshot below
image
- In the sidebar, you will see all newly added controls
image

- Following newly added controls should work as expected:
   - Styles: Fill & Outline ( Feature plugin )
   - Width Settings ( Feature Plugin +  WooCommerce Core )
   - Typography controls ( Feature plugin )

- Also, `block.tsx` file is imported by Cross-sells products [here](https://github.com/woocommerce/woocommerce-blocks/blob/192bdde6c4812dbe626a6444a37a6897d159f2d1/assets/js/blocks/cart/cart-cross-sells-product-list/cart-cross-sells-product.tsx#L18-L19). So make sure that Add to cart button still works fine for Cross-sell products. You can find more about cross-sell products [here](https://woocommerce.com/document/related-products-up-sells-and-cross-sells/).

**Known issues**
- Although all of new controls should also work for Add to cart button in `All products` block, but due to already reported issue [#7930](https://github.com/woocommerce/woocommerce-blocks/issues/7930), these won't work for now. IMO, This issue should be fixed when we will work on [#7930](https://github.com/woocommerce/woocommerce-blocks/issues/7930) issue. What do you think?

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [X] WooCommerce Core
* [X] Feature plugin
* [ ] Experimental

### Changelog

> Add Fill & Outline styles, width settings & new typography controls for Product(Add to cart) button block.

### CODE REVIEW

**Comment by imanish003** [comment]
Hi @Aljullu, I requested a review from you too, because following will require your opinion too:

> Known issues
Although all of new controls should also work for Add to cart button in All products block, but due to already reported issue https://github.com/woocommerce/woocommerce-blocks/issues/7930, these won't work for now. IMO, This issue should be fixed when we will work on https://github.com/woocommerce/woocommerce-blocks/issues/7930 issue. What do you think?

Please find more details in PR description 🙏🏻

**Comment by kmanijak** [comment]
Great work @imanish003!

I tested the solution and I found these two discrepancies:

#### Default button (without outline) is shorter on the frontend.
In Editor default and outlined button has the same height. Outlined button in the frontend as well. But default button on the frontend is shorter (smaller padding)

https://user-images.githubusercontent.com/20098064/226564148-3a7f753c-5bd1-4dc1-97f2-14c33e9f048b.mov

#### 100% width button that is technically an anchor exceeds the container (frontend)
"Select Options" or "View products" for variable and grouped products don't respect the container max width
image

Also, for small width (25%) they are wider

image

**EDIT: The issue with width of anchor occurs only in Product Catalog, but not in a post**

---

> Although all of new controls should also work for Add to cart button in All products block, but due to already reported issue https://github.com/woocommerce/woocommerce-blocks/issues/7930, these won't work for now. IMO, This issue should be fixed when we will work on https://github.com/woocommerce/woocommerce-blocks/issues/7930 issue. What do you think?

My opinion is we skip that and focus on Products block, so I agree with you ☝️

File: src/BlockTypes/ProductButton.php [hunk 1]
```diff
@@ -89,8 +89,10 @@ protected function render( $attributes, $content, $block ) {
 			$ajax_add_to_cart_enabled      = get_option( 'woocommerce_enable_ajax_add_to_cart' ) === 'yes';
 			$is_ajax_button                = $ajax_add_to_cart_enabled && ! $cart_redirect_after_add && $product->supports( 'ajax_add_to_cart' ) && $product->is_purchasable() && $product->is_in_stock();
 			$html_element                  = $is_ajax_button ? 'button' : 'a';
-			$styles_and_classes            = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes, array( 'border_radius', 'font_size', 'font_weight', 'margin', 'padding', 'text_color' ) );
+			$styles_and_classes            = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes );
 			$text_align_styles_and_classes = StyleAttributesUtils::get_text_align_class_and_style( $attributes );
+			$classname                     = isset( $attributes['className'] ) ? $attributes['className'] : '';
+			$custom_width_classes          = isset( $attributes['width'] ) ? 'has-custom-width wp-block-button__width-' . $attributes['width'] : '';
```
Comment (kmanijak) [inline src/BlockTypes/ProductButton.php:95]: What do you think about extending the `get_classes_and_styles_by_attributes` method with the `custom_width_classes` so it's there by default?

I did that for `text-align` attribute in this PR:
https://github.com/woocommerce/woocommerce-blocks/pull/8248/files#diff-bf7a423d751aba56975f9ca86523d396e462e59edf9d9d572109cad8617356c8

Or at least move the CSS class extraction to the `StyleAttributesUtils` class so it's shared and available for other blocks as well?

**Comment by Aljullu** [comment]
> - Although all of new controls should also work for Add to cart button in `All products` block, but due to already reported issue [#7930](https://github.com/woocommerce/woocommerce-blocks/issues/7930), these won't work for now.

That's weird, I'm no longer able to reproduce #7930 (I tested changing the color of the Add to Cart button):

Editor | Frontend
--- | ---
imatge | imatge

> IMO, This issue should be fixed when we will work on [#7930](https://github.com/woocommerce/woocommerce-blocks/issues/7930) issue. What do you think?

I don't think I'm entitled to make a decision on this. I agree that the priority is the Products block, but at the same time the All Products block is widely used so I'm not sure about having an experience that feels broken. Tagging @danielwrobert for his thoughts on this.

**tl;dr** when adding support for new styles to the _Products_ block product element blocks, they are exposed to the _All Products_ block product element blocks as well. To make these new styling options work in product element blocks inside _All Products_ block, extra work is required.

Options we have:

1. Don't fix it, as the priority right now is the Products block, we shouldn't spend time in the All Products block.
2. Fix it, as the All Products block is widely used and it hasn't been officially deprecated yet.
3. Somewhere in between: time-box some hours to see if it's easy to fix it in the All Products block. If it's easy, let's fix it, otherwise, let's leave it as-is and continue the work with the Products block.

(I lean towards option 3, but that depends on whether it's worth continuing spending time on the All Products block)

**Comment by danielwrobert** [comment]
@Aljullu thank you for the ping and for the TL;DR - that's super helpful! 🍻

I agree with option 3. While we don’t want to leave the experience broken for folks who are currently using the All Products block, the Products Block is meant to replace it and we don’t want to let the former hinder the progress of the latter. Time-boxing to determine the necessary effort seems like a good balance.

@imanish003 since we already have an open Issue in #7930 it seems like we’re planning to address this either way - it’s just a matter of _when_ – i.e., do we want to hold up the progress of this PR or, if it requires a more involved fix, merge it and circle back. Is that correct?

**Comment by imanish003** [comment]
Hi @Aljullu @danielwrobert, Thank you for sharing your thoughts 🙏🏻

> That's weird, I'm no longer able to reproduce https://github.com/woocommerce/woocommerce-blocks/issues/7930 (I tested changing the color of the Add to Cart button):

I also tested again, and it looks like all the controls are working fine **except `Width setting`**, which will be added in this PR. As this is something which will be introduced in this PR, therefore it makes sense that this should be fixed in this PR itself. I will look into it 🙂

Also, I think we may close #7930 if we can't reproduce it anymore.

**Comment by Aljullu** [comment]
Thanks!

> Also, I think we may close https://github.com/woocommerce/woocommerce-blocks/issues/7930 if we can't reproduce it anymore.

I went ahead and closed it. :slightly_smiling_face: 

Edit: ups, by mistake I also closed this PR. I have reopened it. :sweat_smile:

File: src/BlockTypes/ProductButton.php (hunk 1 above)
Comment (imanish003) [reply src/BlockTypes/ProductButton.php:95]: Hi Karol! Thanks for pointing this out 🙌🏻

I am not 100% sure about this. Because this is a custom implementation for Add to cart button for now. You can see the code [here](https://github.com/woocommerce/woocommerce-blocks/blob/60c23add44c8bd2e2d17ebbf8f6f402fc94354ff/assets/js/atomic/blocks/product-elements/button/edit.tsx#L28-L72). Therefore, this functionality isn't shared with any other block for now.

But in future, If we will abstract this into a reusable block, then I agree with you that we will need to extend `get_classes_and_styles_by_attributes`.

What do you think?

**Comment by imanish003** [comment]
Hi, @kmanijak 👋  I have made the requested changes except [this one](https://github.com/woocommerce/woocommerce-blocks/pull/8781#discussion_r1143114994), in which I have replied to your suggestion for further discussion 🙂

Please let me know if you see any issues, except `Width setting` not working in All products block.

File: src/BlockTypes/ProductButton.php (hunk 1 above)
Comment (kmanijak) [reply src/BlockTypes/ProductButton.php:95]: Hmm.. 🤔 Makes sense to me, thanks for explanation!

**Comment by imanish003** [comment]
@Aljullu @kmanijak I have made changes in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/8781/commits/417e19612db8aed8257d373cec99acb4b7794e4f) to support Width Settings in `All Products` block. Can you please have a look? Because I am not 100% if that's the right way to do it. 🙏🏻

**Comment by kmanijak** [comment]
> Hi, @kmanijak 👋 I have made the requested changes except https://github.com/woocommerce/woocommerce-blocks/pull/8781#discussion_r1143114994, in which I have replied to your suggestion for further discussion 🙂
> Please let me know if you see any issues, except Width setting not working in All products block.

> @Aljullu @kmanijak I have made changes in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/8781/commits/417e19612db8aed8257d373cec99acb4b7794e4f) to support Width Settings in All Products block. Can you please have a look? Because I am not 100% if that's the right way to do it. 🙏🏻

I retested and the issues I reported earlier are gone 💪 However, neither width nor outline is applied to the All Products on the frontend (they work fine in Editor).

EDIT:
After more testing:
- Styles are applied correctly in the "Edit Mode" of All Products
- Only Width is applied correctly in the Editor in "Preview Mode" - outline is ignored
- Width is applied on the frontend from previous "Save" (e.g. if I change settings and save three times, what I see in frontend is the result of second save, not the last third one). Outline is ignored

**Comment by imanish003** [comment]
@kmanijak I fixed the issue in [ffa2c46](https://github.com/woocommerce/woocommerce-blocks/pull/8781/commits/ffa2c460f3384acf98fbcd2b41fcdcca4a5ed363). Thanks for pointing this out. 🙌🏻

File: assets/js/atomic/blocks/product-elements/button/index.ts [hunk 2]
```diff
@@ -32,6 +34,11 @@ const blockConfig: BlockConfiguration = {
 	attributes,
 	supports,
 	edit,
+	save,
+	styles: [
+		{ name: 'fill', label: 'Fill', isDefault: true },
+		{ name: 'outline', label: 'Outline' },
```
Comment (kmanijak) [inline assets/js/atomic/blocks/product-elements/button/index.ts]: I think we could translate labels here

**Review by kmanijak — APPROVED** [review]
Works like a charm in Products and All Products blocks! 💪 

I left one small comment, so I'm pre-approving ✅ 

One question: you marked PR as a "Feature Plugin" but I don't see the width setting will enter the WC Core if I'm not mistaken. In that case I think it's better to mark "WC Core", what do you think?

**Comment by imanish003** [comment]
Hi @kmanijak, Thanks for pointing this out 🙌🏻

> One question: you marked PR as a "Feature Plugin" but I don't see the width setting will enter the WC Core if I'm not mistaken. In that case I think it's better to mark "WC Core", what do you think?

You are right, As `Width setting` will be included in WC core too so I am checking both `Feature plugin` & `WooCommerce Core` checkbox for `WooCommerce Visibility`. 

Also, I have updated testing step to include it i.e.:
```
Styles: Fill & Outline ( Feature plugin )
Width Settings ( Feature Plugin + WooCommerce Core )
Typography controls ( Feature plugin )
```

File: assets/js/atomic/blocks/product-elements/button/index.ts (hunk 2 above)
Comment (imanish003) [reply assets/js/atomic/blocks/product-elements/button/index.ts]: Nice catch! I have added translation for labels in [2cd7e66](https://github.com/woocommerce/woocommerce-blocks/pull/8781/commits/2cd7e661158b31ff25e6e9987f511a77ea5e4d95). Can you please have a quick look? 🙂
